New brunch with Task WebDriver

yauhen07 opened the pull request.

Commit:
New brunch with Task WebDriver

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+public class NewAppTest extends BaseTest {
+
+    @Test
+    public void newApplication() {
```
recommended: usually test cases have move descriptiva names like 
createNewApplicationWithDefaultSetup()

yauhen07 replied:
Done. If you mean more. Or you want me to move something to somewhere?

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Test
+    public void newApplication() {
+        int x = (int) (Math.random() * 1000);
```
creation of data and all not directly connected with test flow usually moved to another helper method
public getRandomInteger(){
....
}

yauhen07 replied:
I'm planing to it in next home task. You mean move it to another method in current class or in separate class?

evgeny-tumashchick replied:
any option better in separate class in util package
RandomValueGenerator class
make method static

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+
+public class NewAppTest extends BaseTest {
+
+    @Test
```
Test has also a description where u can add more info

yauhen07 replied:
I know, thank you

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+        int x = (int) (Math.random() * 1000);
+        driver.get("https://network-international--qa.cs83.my.salesforce.com/");
+        System.out.println("Open application (url) for testing");
+        Assert.assertEquals(driver.getTitle(), "Login | Salesforce", "Incorrect page (login page) is not opened");
```
Incorrect page (login page) is opened or login page is not opened

yauhen07 replied:
Done

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+        driver.get("https://network-international--qa.cs83.my.salesforce.com/");
+        System.out.println("Open application (url) for testing");
+        Assert.assertEquals(driver.getTitle(), "Login | Salesforce", "Incorrect page (login page) is not opened");
+        System.out.println("Check that correct Login page is opened");
```
u now use sout for logging but later it would be logger.
Overall information like this logically should go before verification -> Check that correct Login page is opened

yauhen07 replied:
Done

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+        Assert.assertEquals(driver.getTitle(), "Login | Salesforce", "Incorrect page (login page) is not opened");
+        System.out.println("Check that correct Login page is opened");
+        WebElement username = driver.findElementByXPath("//input[@class='input r4 wide mb16 mt8 username']");
+        username.sendKeys("yauhen_valodzin@epam.com.qa");
```
there are cases when it could be some data in field so better first clear and then input

evgeny-tumashchick replied:
as clear and input are two steps u can create a separate method for imputing data in the field
public inputData(WebElement el) {
el.clear()
el.sendKeys()
}

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+        Assert.assertEquals(driver.getTitle(), "Login | Salesforce", "Incorrect page (login page) is not opened");
+        System.out.println("Check that correct Login page is opened");
+        WebElement username = driver.findElementByXPath("//input[@class='input r4 wide mb16 mt8 username']");
+        username.sendKeys("yauhen_valodzin@epam.com.qa");
```
move all constants like url u open, login and passwd ant other to constants in test class
e.g.
public final static SALESFORCE_URL = "https://network-international--qa.cs83.my.salesforce.com/";

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+        System.out.println("Open application (url) for testing");
+        Assert.assertEquals(driver.getTitle(), "Login | Salesforce", "Incorrect page (login page) is not opened");
+        System.out.println("Check that correct Login page is opened");
+        WebElement username = driver.findElementByXPath("//input[@class='input r4 wide mb16 mt8 username']");
```
in variable names use the type of element u use
nameInput
applyButton
...

yauhen07 replied:
Done

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+        WebElement button = driver.findElementByXPath("//input[@class='button r4 wide primary']");
+        button.click();
+        System.out.println("Find and click Login button");
+        WebElement dynamicElement = (new WebDriverWait(driver, 10).until(ExpectedConditions.presenceOfElementLocated((By.xpath("//span[text()='Recently Viewed']")))));
```
do we need to save it in dynamicElement. It will be not used later

yauhen07 replied:
Done

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+        System.out.println("Find and click Login button");
+        WebElement dynamicElement = (new WebDriverWait(driver, 10).until(ExpectedConditions.presenceOfElementLocated((By.xpath("//span[text()='Recently Viewed']")))));
+        System.out.println("Add explicit wait");
+        Assert.assertEquals(driver.getTitle(), "Recently Viewed | Applications | Salesforce", "Incorrect Home page is not opened");
```
correct the assertion message

yauhen07 replied:
Whats wrong with it?

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+        formNumberOfAdditionalOutlets.sendKeys("1");
+        System.out.println("Find and populate Number of additional outlets field");
+        WebElement formBusinessNature = driver.findElementByXPath("//div/select/option[@value='Computer Services']");
+        formBusinessNature.click();
```
add some verification after click that action applied. And following actions also

yauhen07 replied:
I'm not sure, should I add verification after each action (input, click) in perfect world?

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewAppTest.java:
```diff
@@ -0,0 +1,72 @@
+        System.out.println("Find and select value from Business Nature dropdown");
+        WebElement formButtonNext = driver.findElementByXPath("//button[text()='Next']");
+        formButtonNext.click();
+        System.out.println("Find and click Next button");
```
as I understand this is a new screen here so verify it is displayed

yauhen07 replied:
Done

Commit:
Add updated gitignore to root folder

Commit:
ATM2018Q2 - WebDriver Basics and Locators for Java
Updated according Mentor's comments - Second try

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/BaseTest.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.concurrent.TimeUnit;
+
+public class BaseTest {
+    ChromeDriver driver;
```
make protected

yauhen07 replied:
Done

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewApplicationTest.java:
```diff
@@ -0,0 +1,74 @@
+    @Test
+    public void createNewApplicationWithDefaultSetup() {
+        int x = (int) (Math.random() * 1000);
+        driver.get("https://network-international--qa.cs83.my.salesforce.com/");
```
move url login pass and other hardcoded text to constants

yauhen07 replied:
Done

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewApplicationTest.java:
```diff
@@ -0,0 +1,74 @@
+        System.out.println("Add explicit wait");
+        new WebDriverWait(driver, 10).until(ExpectedConditions.presenceOfElementLocated((By.xpath("//span[text()='Recently Viewed']"))));
+        System.out.println("Check that correct Home page (after login) is opened");
+        Assert.assertEquals(driver.getTitle(), "Recently Viewed | Applications | Salesforce", "Incorrect Home page is not opened");
```
Move expected text results to constants also

yauhen07 replied:
Done

evgeny-tumashchick commented on ATM_Selenium_Training/src/test/java/com/epam/atm/selenium/NewApplicationTest.java:
```diff
@@ -0,0 +1,74 @@
+        System.out.println("Find and open Business Line drop down");
+        WebElement formBusinessLineValue = driver.findElementByXPath("//a[text()='Computer Network/Information Services']");
+        formBusinessLineValue.click();
+        System.out.println("Find and select value from Business drop down");
```
add assertion that Computer Network/Information Services text is displayed

yauhen07 replied:
Done

Commit:
ATM2018Q2 - WebDriver Basics and Locators for Java
Updated according Mentor's comments - Third try
Updated sctructure of the project

Commit:
ATM2018Q2 - WebDriver Basics and Locators for Java
Update constants

evgeny-tumashchick reviewed (approved):
Please remove empty file before merge

Commit:
ATM2018Q2 - WebDriver Basics and Locators for Java
Remove temp ddd.java file
Update assert
